### ISSUE TITLE
numpy.sum not stable enough sometimes (Kahan, math.fsum)

### ISSUE THREAD

**Issue by nschloe** [numpy/numpy#8786]
#### Background

This came up developing [quadpy](https://github.com/nschloe/quadpy) for numerical quadrature. To approximate an integral over the reference interval `[-1, 1]`, one arranges points `x_i` and weights `w_i` such that the number
```
sum(f(x_i) * w_i)
```
approximates the integral of some functions `f` (typically a set of polynomials) reasonably well. Sophisticated schemes have many points and weights, and naturally the more weights there are, the smaller each individual weight must be.

*Bottom line: You'll have to sum up many small values.*

#### When numpy.sum doesn't cut it

I noticed that `numpy.sum` produces large enough round-off errors for the tests to fail with as few as 7 summands (being `-44.34121805, -15.72356145, -0.04563372, 0., 0.04563372, 15.72356145, 44.34121805`, not shown in full precision). `math.fsum` always worked fine, but cannot be used anymore since I'd like to compute the sum for many intervals at once.

#### Work so far

[Kahan's summation](https://en.wikipedia.org/wiki/Kahan_summation_algorithm)
```python
def kahan_sum(a, axis=0):
    s = numpy.zeros(a.shape[:axis] + a.shape[axis+1:])
    c = numpy.zeros(s.shape)
    for i in range(a.shape[axis]):
        # http://stackoverflow.com/a/42817610/353337
        y = a[(slice(None),) * axis + (i,)] - c
        t = s + y
        c = (t - s) - y
        s = t.copy()
    return s
```
does the trick. Note, however, the use of a Python loop here, which leads to a performance degradation if the number of points and weights is large.

`math.fsum` [uses a different algorithm](https://github.com/python/cpython/blob/d267006f18592165ed97e0a9c2494d3bce25fc2b/Modules/mathmodule.c#L1087) which may or may not be better suited than Kahan.

**Comment by eric-wieser** [comment]
I don't think that `copy` is needed there.

These seems like an improvement that could be built directly in to the `sum` function if #8773 were implemented

**Comment by juliantaylor** [comment]
That might be the problem, the fast axis in numpy is the last one by default (C order).
How does it fare when you have your arrays in fortran order?

**Comment by eric-wieser** [comment]
> I see the same deficiencies.

Yes, that's because you're doing exactly the same operation over memory, and just calling the axes different things

`np.copy(c, order='f').sum(axis=0)` is the relevent test.

**Comment by seberg** [comment]
You would have to sparkle in a `copy` there after the transpose, or do `copy("F")` and not transpose.

**Comment by njsmith** [comment]
Another quick hack to get better accuracy is to sort your array before summing.

It'd certainly be nice to have better summation algorithms available as an option, but `np.sum` is unlikely to ever use them by default given the performance cost.

**Comment by seberg** [comment]
Right, silly by me. The pairwise stuff could maybe actually have helped,since it might sum all the negative and then all the positive ones one after another (the example actually was sorted, but basically symmetric about zero, which invites anihilation)?

**Comment by nschloe** [comment]
>  `np.sum` is unlikely to ever use them by default given the performance cost.

Definitely not, I agree.

It'd be nice to have more options, though, for example `numpy.kahan_sum` with the same signature as `numpy.sum`.

**Comment by charris** [comment]
The same algorithm used to generate a Hamming code could also be adapted: always summing the two numbers of smallest absolute value. Could be implemented as a priority queue. I'm not recommending that, have just wondered now and then how well it would work...

**Comment by njsmith** [comment]
Sorry, yeah, you would want to sort by absolute value. (I've only used this trick for summing probabilities, where this doesn't come up, so I forgot :-).)

**Comment by nschloe** [comment]
> always summing the two numbers of smallest absolute value.

You can do this when summing `float`s, but not when summing arrays.

**Comment by njsmith** [comment]
@nschloe: you have to sort each array separately, yes. It's awkward but certainly possible, as an option.

**Comment by wuisawesome** [comment]
Would either a kahan_sum, fsum, or both be a feature others think is worth implementing?

**Comment by asmeurer** [comment]
A stable numpy sum would be awesome. Sorting is a PITA since you have to do it by a key (abs), and if you have a more than 1-d array you have to sort each subarray independently.

**Comment by cdeil** [comment]
We would also like to have a more precise sum, for a likelihood that's the sum of ~ a billion numbers.

@nschloe - Very nice comparison in `accupy`, thanks!

So your conclusion is that the precision of kahan sum isn't much better, so not worth exposing in Numpy?

(kahan is actually already [here](https://github.com/numpy/numpy/blob/bc66801a9e578efb73c73473ee7a99748c3214f1/numpy/random/common.pyx#L173-L183) for a different application, not for numpy.sum)

fsum accuracy seems amazing. How slow is it?
The speed comparison in accupy isn't fair since it's using a Python for loop, no?
(See [here](https://github.com/nschloe/accupy/blob/63a031cab7f4d3b9ba1073f9328c10c1862d1c4d/accupy/sums.py#L72)).
Is a fsum available (in C or Cython or Numba or ...) somewhere that works on Numpy arrays?

**Comment by seberg** [comment]
@cdeil python uses Shewchuk summation, my guess is you can just look at their code or google a C implementation at the very least. Adapting it into short cython code is probably less work then trying to dig very long.

Now that we support the axis argument for gufuncs, it could actually be a nice little exercise to implement a gufunc for stable summation. Not sure it will be ideal performance wise, but cython solution are unlikely to be any better (unless they bloat up a lot).

**Comment by nschloe** [comment]
> So your conclusion is that the precision of kahan sum isn't much better, so not worth exposing in Numpy?

Not my conclusion but a known property of Kahan. It's just [not robust when it comes to ill-conditioned sums](https://en.wikipedia.org/wiki/Kahan_summation_algorithm#Accuracy). The best thing would be have fsum indeed, but there are faster solutions too which work in k-fold machine precision. (See [accupy](https://github.com/nschloe/accupy).)

**Comment by lorentzenchr** [comment]
To add a simple example:
```python
import numpy as np

def naive_sum(a):
    result = np.zeros_like(a, shape=(1,))
    for i in range(a.shape[0]):
        result[0] += a[i]
    return result[0]

a = np.linspace(-100, 100, 1_000_000, dtype=np.float32)
np.random.default_rng(42).shuffle(a)

sum(a), naive_sum(a), np.sum(a)
# (0.0, -0.5259266, -0.008300781)
```
This shows that `numpy.sum` does better than the naive summation.

And for reference, Python's `sum` meanwhile uses (Kahan-) Neumaier summation, see https://github.com/python/cpython/issues/100425.

**Comment by nschloe** [comment]
For those interested, here is an example with a badly conditioned sum:

```python
import math
import numpy as np

# An ill-conditioned sum, condition number ~2.188e+14
p = [
    -0.41253261766461263,
    41287272281118.43,
    -1.4727977348624173e-14,
    5670.3302557520055,
    2.119245229045646e-11,
    -0.003679264134906428,
    -6.892634568678797e-14,
    -0.0006984744181630712,
    -4054136.048352595,
    -1003.101760720037,
    -1.4436349910427172e-17,
    -41287268231649.57,
]

exact = -0.377392919181026

s = math.fsum(p)
print("math.fsum()", s, abs(s - exact))

s = sum(p)
print("sum():      ", s, abs(s - exact))

s = np.sum(p)
print("np.sum():   ", s, abs(s - exact))
```
```
math.fsum(): -0.377392919181026 0.0
sum():       -0.375 0.002392919181026021
np.sum():    -0.3828125 0.005419580818973979
```

`fsum()` computes the sum without error, `sum()` and `np.sum()` are still in the error order of magnitude 10^{-3}.

**Comment by asmaier** [comment]
Using Python 3.12.3 on Mac OS 14.1.1 with Apple M3 Pro I get the following results from the script by @nschloe: 

```
math.fsum() -0.377392919181026 0.0
sum():       -0.377392919181026 0.0
np.sum():    -0.3828125 0.005419580818973979
naive_sum():    -41287268231649.57 41287268231649.195
```
The reason why `math.fsum()` and `sum()` now yield the same result might be that since Python 3.12 the algorithm for `sum()` was improved: "sum() now uses Neumaier summation to improve accuracy and commutativity when summing floats or mixed ints and floats. ", see [docs.python.org/3/whatsnew/3.12.html](https://docs.python.org/3/whatsnew/3.12.html)

**Comment by NimaSarajpoor** [comment]
I am not sure whether I should open a new issue or not. Please let me know. 


**Issue:**
In the following example, I've created two 1D numpy arrays. Both have the same non-zero elements but at different positions. They have different sum !


**Example:**
```
import numpy as np

non_zero_elems = """
0.024861743494183196  0.024861743494183196  0.024861743494183196
 0.021792470064670937  0.021792470064670937  0.018949939845733486
 0.018949939845733486  0.0015757541842979532 0.0015757541842979532
 0.0015757541842979532 0.0015757541842979532 0.0015757541842979532
 0.0015757541842979532 0.0015757541842979532 0.0015757541842979532
 0.04033997886552635   0.04033997886552635   0.04033997886552635
 0.04033997886552635  
 """
non_zero_elems = np.array(non_zero_elems.split()).astype(np.float64)
IDX1 = np.array([28, 29, 30, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52,
       53, 54])

IDX2 = np.array([ 8,  9, 10, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52,
       53, 54])

# `x` and `y` are two arrays with the same non-zero elements but at different positions
x = np.zeros(55, dtype=np.float64)
y = np.zeros(55, dtype=np.float64)

x[IDX1] = non_zero_elems.copy()
y[IDX2] = non_zero_elems.copy()

print(np.sum(x) == np.sum(y))  # False!! 
print(np.sum(x[x!=0]) == np.sum(y[y!=0]))  # True
```

**Comment by seberg** [comment]
> In the following example, I've created two 1D numpy arrays. Both have the same non-zero elements but at different positions. They have different sum !

It's not an issue, it is a basic property of numerical floating point math that re-ordering even the simplest operations can lead to different results.  More stable summation may do that less, but I don't think there is any reasonable version that is completely exact.

**Comment by lorentzenchr** [comment]
@seberg What do numpy maintainers intend to do about this issue. I see several options:
1. Close the issue as a "non issue" or "won't be implemented".
2. Add an option to the existing `sum` (what about `mean` and `average`?)
3. Add a new function `numpy.fsum` that simply promotes python's `fsum` to a ufunc.
4. Do nothing.
5. Yet another option.

**Comment by seberg** [comment]
Well, I think there is nothing to do about mean/sum directly.  We could think about a new function or maybe about a new option to sum/mean (but that is harder).
So I think it's a reasonable request, but a hard one.  A new function would be the simplest thing probably.
(I don't see anything to do about documentation.)
